text/0240-app-telemetry-schema-vision-roadmap.md

```diff
@@ -0,0 +1,271 @@
+will tend to produce more complex schema hierarchies). A single Telemetry Schema
+can import from zero to n Semantic Convention Registries, depending on its
+needs. A Resolved Telemetry Schema is generated during the schema resolution
+process from a hierarchy of Telemetry Schemas and Semantic Convention
+Registries.
+
+The following diagram illustrates a possible instance of a complex hierarchy of
+schemas and semantic convention registries. It involves several vendor and
+enterprise artifacts, in addition to the standardized OpenTelemetry artifacts.
+The schema resolution process will produce a self-contained Resolved Telemetry
+Schema that can be easily consumed by various tools and applications, such as a
+Client SDK generator, compatibility checker, compliance checker, data catalog
+feeder, and more.
+
+![Example of Application Telemetry Schema Hierarchy](./img/0240-app-telemetry-schema-hierarchy-example.png)
+
+For each important component, the following diagram defines the responsibilities
+and key design properties that have been considered in the subsequent sections
+of this document.
+
+![Application Telemetry Schema - Levels of responsibilities & Key design properties](./img/0240-app-telemetry-schema-design-properties.png)
+
+This design enables the definition of semantic conventions in a distributed
+manner. OpenTelemetry, vendors, and enterprises can define their own semantic
+conventions in different registries simplifying the existing process.
+
+The next two sections will describe the concepts of Telemetry Schema v1.2 and
+the Resolved Telemetry Schema.
+
+## Telemetry Schema v1.2 (aka Application Telemetry Schema)
+
+The Telemetry Schema v1.2 represents a forward-compatible extension of the
+existing Telemetry Schema v1.1. It offers a way to import semantic convention
+registries and to define resource attributes, the properties of the
+instrumentation library, and the telemetry signals an application or library can
+produce. The existing versioning mechanism defined in version 1.1 remains
+unchanged. A Telemetry Schema can inherit from one or several other telemetry
+schemas (as explained in the previous section). The root schema is usually the
+official Telemetry Schema, referencing the official OpenTelemetry Semantic
+Convention Registry. The last schema (leaf) in this hierarchy is called the
+Application Telemetry Schema, as it defines all the signals a particular
+application produces. Although there is no direct lineage between these systems,
+a similar approach was designed and deployed by Facebook to address the same
+type of problem but in a proprietary context (refer to this
+[positional paper](https://research.facebook.com/publications/positional-paper-schema-first-application-telemetry/)
+for more information).
+
+The following diagram illustrates the relationships between these concepts and
+the main changes from Telemetry Schema v1.1 to v1.2.
+
+![Telemetry Schema](./img/0240-telemetry-schema-v1.1-to-v1.2.png)
```

Comment:
A question regarding this diagram: it shows resources and signals are defined as "siblings" of versions. So if I understand correctly, the schema defines resources and signals for the latest version in the schema, and then additional versioned transformations to transform attributes of older versions.

I wonder if there's value in defining those resources and signals for every version, as also those definitions can change across versions? Or asked in a different way: do we want to be able to create SDKs for each version contained in the schema file?

Reply:
For simplicity's sake, I was thinking of following the principles adopted by protobuf (forward-/backward compatibility rules) in terms of signals and resources, with the slight difference that the elements defined in the semantic convention registries (mainly attribute and metric definitions) follow the rules of transformation of the telemetry schema v1.1. Therefore, by default, a client SDK will be generated for the most recent version mentioned in the schema. However, it will be entirely possible for a backend component to consume data corresponding to an earlier version of the attributes and metrics, although in terms of signals and resources, apart from these elements, it will be necessary to be able to support the rules of classical forward-backward compatibility. For the moment, this seems to me a good balance between user experience and adaptability. This topic is clearly a debate to be had and experimented with.
It would be interesting to have @yurishkuro's feedback on this point, as a similar approach has been implemented at Meta. I am due to discuss this with him soon.